pkg/asset/kubeconfig/kubeconfig_test.go

```diff
@@ -0,0 +1,229 @@
+preferences: {}
+users:
+- name: admin
+  user:
+    client-certificate-data: VEhJUyBJUyBBRE1JTiBDRVJUIERBVEE=
+    client-key-data: VEhJUyBJUyBBRE1JTiBLRVkgREFUQQ==
+`),
+		},
+		{
+			userName: "kubelet",
+			certKey:  kubeletCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:         rootCAState,
+				kubeletCertKey: kubeletCertState,
+				installConfig:  installConfigState,
+			},
+			errString: "",
+			expectedData: []byte(`clusters:
+- cluster:
+    certificate-authority-data: VEhJUyBJUyBST09UIENBIENFUlQgREFUQQ==
+    server: https://test-cluster-name-api.test.example.com:6443
+  name: test-cluster-name
+contexts:
+- context:
+    cluster: test-cluster-name
+    user: kubelet
+  name: kubelet
+current-context: kubelet
+preferences: {}
+users:
+- name: kubelet
+  user:
+    client-certificate-data: VEhJUyBJUyBLVUJFTEVUIENFUlQgREFUQQ==
+    client-key-data: VEhJUyBJUyBLVUJFTEVUIEtFWSBEQVRB
+`),
+		},
+		{
+			userName: "admin", // No root ca in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				adminCertKey:  adminCertState,
+				installConfig: installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "admin", // No admin cert in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:         rootCAState,
+				kubeletCertKey: kubeletCertState,
+				installConfig:  installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "kubelet", // No kubelet cert in parents.
+			certKey:  kubeletCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:        rootCAState,
+				adminCertKey:  adminCertState,
+				installConfig: installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "admin", // No install config in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:       rootCAState,
+				adminCertKey: adminCertState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+	}
+
+	for i, tt := range tests {
+		kubeconfig := &Kubeconfig{
+			rootDir:       testDir,
+			userName:      tt.userName,
+			rootCA:        rootCA,
+			certKey:       tt.certKey,
+			installConfig: installConfig,
+		}
+		st, err := kubeconfig.Generate(tt.parents)
+		if err != nil {
+			assert.EqualErrorf(t, err, tt.errString, fmt.Sprintf("test #%d expected %v, saw %v", i, tt.errString, err))
+			continue
+		} else {
+			if tt.errString != "" {
+				t.Errorf("test #%d expect error %v, saw nil", i, tt.errString)
+			}
+		}
+
+		filename := filepath.Join(testDir, "auth", fmt.Sprintf("kubeconfig-%s", tt.userName))
+		data, err := ioutil.ReadFile(filename)
+		if err != nil {
+			t.Errorf("test #%d failed to read kubeconfig %q: %v", i, filename, err)
```

Comment:
nit: Instead of tracking `i`, you can move your explanatory comments into `tt.name` (with the username in `tt.username`?) and `t.Run` the subtests.  For example, see [here][1].

[1]: https://github.com/openshift/release/commit/e1f993fb1debc43d81f7c413cd4fea39077d48b4#diff-233b4d8888505234561dc29b3e905ad5R156